js/src/miscellaneous/resolve.js

```diff
@@ -0,0 +1,112 @@
+/* eslint-env mocha */
+'use strict'
+
+const isIpfs = require('is-ipfs')
+const loadFixture = require('aegir/fixtures')
+const { getDescribe, getIt, expect } = require('../utils/mocha')
+
+module.exports = (createCommon, options) => {
+  const describe = getDescribe(options)
+  const it = getIt(options)
+  const common = createCommon()
+
+  describe('.resolve', () => {
+    let ipfs
+
+    before(function (done) {
+      // CI takes longer to instantiate the daemon, so we need to increase the
+      // timeout for the before step
+      this.timeout(60 * 1000)
+
+      common.setup((err, factory) => {
+        expect(err).to.not.exist()
+        factory.spawnNode((err, node) => {
+          expect(err).to.not.exist()
+          ipfs = node
+          done()
+        })
+      })
+    })
+
+    after((done) => {
+      common.teardown(done)
+    })
+
+    it('should resolve an IPFS hash', (done) => {
+      const content = loadFixture('js/test/fixtures/testfile.txt', 'interface-ipfs-core')
+
+      ipfs.files.add(content, (err, res) => {
+        expect(err).to.not.exist()
+        expect(isIpfs.cid(res[0].hash)).to.be.true()
+
+        ipfs.resolve(`/ipfs/${res[0].hash}`, (err, path) => {
+          expect(err).to.not.exist()
+          expect(path).to.equal(`/ipfs/${res[0].hash}`)
+          done()
+        })
+      })
+    })
+
+    // Test resolve turns /ipfs/QmRootHash/path/to/file into /ipfs/QmFileHash
+    it('should resolve an IPFS path link', (done) => {
+      const path = '/path/to/testfile.txt'
+      const content = loadFixture('js/test/fixtures/testfile.txt', 'interface-ipfs-core')
+
+      ipfs.files.add([{ path, content }], { wrapWithDirectory: true }, (err, res) => {
+        expect(err).to.not.exist()
+
+        const rootHash = res.find(r => r.path === '').hash
+        const fileHash = res.find(r => r.path === path).hash
+
+        ipfs.resolve(`/ipfs/${rootHash}${path}`, (err, path) => {
+          expect(err).to.not.exist()
+          expect(path).to.equal(`/ipfs/${fileHash}`)
+          done()
+        })
+      })
+    })
+
+    it('should not resolve an IPFS path non-link', (done) => {
+      const content = { path: { to: { file: 'hello world' } } }
+      const options = { format: 'dag-cbor', hashAlg: 'sha2-256' }
+
+      ipfs.dag.put(content, options, (err, cid) => {
+        expect(err).to.not.exist()
+
+        const path = `/ipfs/${cid.toBaseEncodedString()}/path/to/file`
+        ipfs.resolve(path, (err, path) => {
+          expect(err).to.exist()
+          expect(err.message).to.equal('found non-link at given path')
+          done()
+        })
+      })
+    })
+
+    // Test resolve turns /ipns/domain.com into /ipfs/QmHash
+    it('should resolve an IPNS DNS link', function (done) {
```

Comment:
This case is a bit worrisome. First we have the reliance on a remote service, secondly we're not actually checking if the output is correct.

My first thought is that in general for the ipns tests, we should have a mini-dns service running together with the tests, that the local daemon can resolve IPNS to. So instead of testing `/ipns/ipfs.io` we should test `/ipns/local-dns-service.home` which would be resolved only locally and we can then assert correct output.

Reply:
I would like that too. I have no idea how to do that or if it's even possible but I'll look into it. Do you have any pointers?